### DISCUSSION TITLE
Add a new data source myself: Snowflake

### DISCUSSION THREAD

**Discussion by sergioluka** [chartbrew/chartbrew#198] [Integration request]
I'd like to use Snowflake as the data source.
Is there a recommended way to extend the tool to add a new source on my side?

**Comment by razvanilin** [comment]
Hey @sergioluka at the moment there's no guide on how to add new data sources, but it's definitely something I want to work on soon.

Currently, it's a bit messy, but here are a few tips on what the change involves:

**Backend:** 
* Add a Snowflake-specific method in the **ConnectionController** class, like this one: https://github.com/chartbrew/chartbrew/blob/master/server/controllers/ConnectionController.js#L619
* Extend new condition here for Snowflake: https://github.com/chartbrew/chartbrew/blob/master/server/controllers/ConnectionController.js#L282
* Add new condition for Snowflake in the **DatasetController**: https://github.com/chartbrew/chartbrew/blob/master/server/controllers/DatasetController.js#L196
* Add new condition for Snowflake in the **DataRequestController**: https://github.com/chartbrew/chartbrew/blob/master/server/controllers/DataRequestController.js#L97

**Frontend:**
* Each new connection will need at least two components, a Builder used to make queries and a Form to add the new data source
* Add **Snowflake** config in the available connections: https://github.com/chartbrew/chartbrew/blob/master/client/src/modules/availableConnections.js
* Add a **Snowflake** image in both light and dark mode here: https://github.com/chartbrew/chartbrew/blob/master/client/src/config/connectionImages.js
* Add a new **Snowflake** folder in the Connections folder. Here is an example with Firestore: https://github.com/chartbrew/chartbrew/tree/master/client/src/containers/Connections/Firestore
* The Form is added in the **ConnectionWizard** so users can select it from the list: https://github.com/chartbrew/chartbrew/blob/master/client/src/containers/Connections/ConnectionWizard.jsx#L296
* The Builder is added in the **DatasetQuery** container: https://github.com/chartbrew/chartbrew/blob/master/client/src/containers/Dataset/DatasetQuery.jsx#L354

I think in the future it will be good to be able to add these as separate modules, but still can't get my head around how it would work when the app has the server and client apps running separately. In the front-end, at least, it should simplify the work a lot

**Reply by sergioluka** [reply]
@razvanilin thank you very much for the detailed answer! Maybe adding functionality to add a custom (DB) connection could b standardized enough to become source-agnostic.
